rclcpp/include/rclcpp/node.hpp

```diff
@@ -258,133 +258,584 @@ class Node : public std::enable_shared_from_this<Node>
+   * \throws rclcpp::exceptions::ParameterNotDeclaredException if any of the
+   *   parameters have not been declared and undeclared parameters are not
+   *   allowed.
+   */
   RCLCPP_PUBLIC
   std::vector<uint8_t>
   get_parameter_types(const std::vector<std::string> & names) const;
 
+  /// Return a list of parameters with any of the given prefixes, up to the given depth.
+  /**
+   * ??? how to list all parameters? what does depth 0 and 1 do?                
+   */
   RCLCPP_PUBLIC
   rcl_interfaces::msg::ListParametersResult
   list_parameters(const std::vector<std::string> & prefixes, uint64_t depth) const;
 
+  using OnParametersSetCallbackType =
+    rclcpp::node_interfaces::NodeParametersInterface::OnParametersSetCallbackType;
+
+  /// Register a callback to be called anytime a parameter is about to be changed.
+  /**
+   * The callback signature is designed to allow handling of any of the above
+   * `set_parameter*` or `declare_parameter*` methods, and so it takes a const
+   * reference to a vector of parameters to be set, and returns an instance of
+   * rcl_interfaces::msg::SetParametersResult to indicate whether or not the
+   * parameter should be set or not, and if not why.
+   *
+   * For an example callback:
+   *
+   *   rcl_interfaces::msg::SetParametersResult
+   *   my_callback(const std::vector<rclcpp::Parameter> & parameters)
+   *   {
+   *     rcl_interfaces::msg::SetParametersResult result;
+   *     result.successful = true;
+   *     for (const auto & parameter : parameters) {
+   *       if (!some_condition) {
+   *         result.successful = false;
+   *         result.reason = "the reason it could not be allowed";
+   *       }
+   *     }
+   *     return result;
+   *   }
+   *
+   * You can see that the SetParametersResult is a boolean flag for success
+   * and an optional reason that can be used in error reporting when it fails.
+   *
+   * This allows the node developer to control which parameters may be changed.
+   *
+   * Note that the callback is called when declare_parameter() and its variants
+   * are called, and so you cannot assume the parameter has been set before
+   * this callback, so when checking a new value against the existing one, you
+   * must account for the case where the parameter is not yet set.
+   *
+   * Some constraints like read_only are enforced before the callback is called.
+   *
+   * There may only be one callback set at a time, so the previously set
+   * callback is returned when this method is used, or nullptr will be returned
+   * if no callback was previously set.
+   *
+   * \param[in] callback The callback to be called when the value for a
+   *   parameter is about to be set.
+   * \return The previous callback that was registered, if there was one,
+   *   otherwise nullptr.
+   */
+  rclcpp::Node::OnParametersSetCallbackType
+  set_on_parameters_set_callback(rclcpp::Node::OnParametersSetCallbackType callback);
```

Comment:
I’d like feedback on this function (which sort of already exists, see `register_param_change_callback` below) that I’m adding/refactoring as part of this read-only parameter refactor.

My main concern that there will be times when two pieces of code that do not know about each other both need to influence this behavior using this callback. The first concrete example of this is that I’d like to use this callback for checking the type of the `use_sim_time` parameter (which is implicitly declared by each node) before setting it, but if I did that then a user could easily override that callback when they set their own and ignore the return value, breaking the behavior of `use_sim_time`.

Alternatives are that I can simply push back callbacks onto a vector and call of them in the order they were added, or I could instead setup a feature where each parameter can have a single callback in addition to the proposed function (one callback for all parameters). I think the cases where you need to have more than one callback for a single parameter are low, but not impossible, so I think the former alternative is preferable (though I'd like a way to also remove them later).

The reason I’m bringing this up now, is that if I want to fix this in the future and use the alternative which keeps a vector of these, then I feel that we’re introducing a function which should basically be deprecated as soon as possible, and that doesn’t sit well with me.

I didn't do this yet, because even though this pr has ballooned into basically a refactor of the parameter's API, I really was trying to have the least impact on existing behavior, but allowing for undeclared parameters (or not), basically touches every part of the API.

Note, I’m interested in what we would like, rather than when we should do it (now or after Dashing). I can figure that out afterwards.

Reply:
> The first concrete example of this is that I’d like to use this callback for checking the type of the use_sim_time parameter (which is implicitly declared by each node) before setting it, but if I did that then a user could easily override that callback when they set their own and ignore the return value, breaking the behavior of use_sim_time.

This doesn't seem too bad at the moment while there is only one parameter `use_sim_time`. Downstream code would have to duplicate validating `use_sim_time` though.

> Alternatives are that I can simply push back callbacks onto a vector and call of them in the order they were added

Are all callbacks given the same parameters? If so, a custom callback that errored when an unknown parameter was given would still need to know about built in parameters like `use_sim_time`.

> or I could instead setup a feature where each parameter can have a single callback in addition to the proposed function (one callback for all parameters)

I like this option with the callback added by the proposed function only getting parameters which don't have a dedicated validation function. I think that allows adding a built in parameter in the future without needing to update a callback set here.

Reply:
We discussed supporting multiple validation callbacks as an enhancement going forward.  It's ticketed in in #296

Reply:
> This doesn't seem too bad at the moment while there is only one parameter use_sim_time. Downstream code would have to duplicate validating use_sim_time though.

I dunno, the first example I write for how to use this callback will already need to be careful to store the return of this function and call it before after the new callback code, so it's pretty bad, imo.

> Are all callbacks given the same parameters? If so, a custom callback that errored when an unknown parameter was given would still need to know about built in parameters like use_sim_time.

Yes, the idea is that you ignore parameters you don't care about.

> I like this option with the callback added by the proposed function only getting parameters which don't have a dedicated validation function. I think that allows adding a built in parameter in the future without needing to update a callback set here.

It wouldn't work like that, because it isn't clear how to handle a case where you get a `set_parameters_atomically({{"p1", ...}, {"p2", ...}})` and `p1` has a dedicated callback, but `p2` doesn't. Do you call `p1`'s callback and the general one, or just `p1's`, or just the general one?

Another reason I favor the former alternative (basically just a vector of callbacks that fire for all parameter set actions) is that it's simpler. You call all the callbacks in order, for each set action unconditionally. The only thing you have to decide is whether to call all and then see if any failed or stop at the first failure (I like the latter better).

Reply:
> We discussed supporting multiple validation callbacks as an enhancement going forward. It's ticketed in in #296

Thanks, I had forgotten about that issue. I suppose I could move the function there, but I wanted to make some headway on _what we want_, so that when we are planning the next release we have something concrete to evaluate.

Also, this brings in a new concern (which my previous response to @rohbotics didn't consider) which is that a single "set verification" callback doesn't make it very easy to have two parts of the code both influence verification without coordination (like with `use_sime_time` and a single user parameter).

So, I'm really asking about that.

Reply:
There are two diverging thoughts I had while reading this thread:

1.  For the specific example you bring up (checking the type and then doing additional verification), it seems to me like checking that the type doesn't change should essentially be the default.  While there are uses for having a parameter that can change types, I think that for the vast majority of cases you just want a parameter to have one type and then error (or really, an ignored update) if someone tries to change that type.  So really, I think the right way to solve this particular example is to introduce type-checking into the low-level callback that happens when a parameter is changed, and then you still only need a single user callback.

1.  With the above being said, I can see that there are uses for having multiple callbacks for a particular parameter that are un-coordinated.  While I think we'll eventually want to end up in a place where we have a vector of callbacks, I'll bring up the point that having multiple callbacks can interact in somewhat unexpected ways for the user and/or the writer of the callbacks.

Reply:
> 1. For the specific example you bring up ...

What about a case where you want to allow int or float? Or boolean or string? I think the user needs to be able to express logic like this if they desire.

However, we can use the meta data about a parameter to check some of these things automatically, like type or immutability or range/choice.

I would prefer to see this kind of checking be the default (as you suggested), but I think it also needs to be possible to opt out of the automatic checking and then do the checking manually.

So the user can, for example, provide a callback which does some custom logic checking and then also calls the verification function that automatically verifies the new value according to the parameter descriptor.

> I'll bring up the point that having multiple callbacks can interact in somewhat unexpected ways for the user and/or the writer of the callbacks.

Yeah, but is there any way around that? Currently if someone sets a callback it's up to them to store and call the old callback if they want. If they overlook this then it will just break things. I don't think that's better or worse than allowing multiple callbacks and failing at the first one (run in the reverse? order of being added).

Most callbacks will only look at a narrow set of parameters or a single parameter. In which case, interference will be minimal.

In the future we can add syntactic sugar that makes it easy for the user to write a callback for a single parameter, getting rid of the boilerplate of checking the parameter name and stuff like that. But to support that and the more advanced use case (a callback looking for a group of parameters or verifying something about all parameters) at the same time we need a vector of callbacks and not a single one, imo.

Reply:
> What about a case where you want to allow int or float? Or boolean or string? I think the user needs to be able to express logic like this if they desire.

I agree, but I also think this is a niche use-case.  Most parameters just need to be a specific type that is settable at runtime.

> Yeah, but is there any way around that?

The basic way around it is to only use a single installed callback.  To be clear, I'm not advocating that we restrict it like it is today; I think that a vector of callbacks makes sense for advanced use-cases.  But I also think that if we get enough "generic" features into the underlying parameter implementation (type checking, read-only, ranges, etc), then most users will only need to bother with 0 or 1 callbacks.

Reply:
>  But I also think that if we get enough "generic" features into the underlying parameter implementation (type checking, read-only, ranges, etc), then most users will only need to bother with 0 or 1 callbacks.

That's a fair point, but I am concerned about designing ourselves into a corner where the niche cases are no longer possible to support without breaking API significantly or having lots of redundant API's.

I also would really like to see us implement the (let's call them) "parameter constraints" using the same mechanism available to the user. So it's transparent (just another callback in the list), composable (user can call from their callback conditionally), and possible to disable the default checks (remove from callbacks list).